### ISSUE TITLE
Pass runtime parameters only to specific implementation 

### ISSUE THREAD

**Issue by elibixby** [fsspec/filesystem_spec#1026]
Config file based options for passing specific arguments to specific backends don't work for non-string parameters. 

For example, from the wording of #453 if I want to pass a custom `google.auth.Credentials` object to `gcs` then
I expected be able to do one of the following:

```
file.open(path, gcs_token=creds)
file.open(path, gcs={'token': creds})
```

But neither work (evident why from looking at the code, there's actually no change to how runtime parameters are parsed, only config files). There doesn't appear to be any way to force runtime parameters that can't be serialized into a json or ini file to be passed to a specific backend.

The only available workaround is something like:
```
fs, _, _ = fsspec.core.get_fs_token_paths(path)
storage_options = {}
if isintance(fs, gcsfs.GCSFileSystem):
   storage_options['token'] = creds
return fsspec.open(path, **storage_options)
```
Essentially building the storage class twice and requiring an explicit check. Eliminating a lot of the utility of fsspec in the first place.

I understand that implementations may have keyword arguments that clash with protocol names, hence not wanting to use the first solution. But it seems reasonable to have another argument, "protocol_defaults" which allows passing a config dictionary directly.
```
fsspec.open(path,  protocol_defaults={'gcs': {'token': creds}})
```

This could be implemented internally with a context manager (also user facing for code cleanliness):

```
with fsspec.config.protocol_defaults(gcs={'token': creds}):
    fsspec.open(path)
```

**Comment by martindurant** [comment]
I agree that it would be a good idea to have a context manager setting temporary config defaults for the duration of the following block. Dask implements something similar. You should be able to modify the config in place, however:
```
import fsspec.config
fsspec.config.conf.setdefault("gcs", {})["token"] = creds
```
Obviously this remains in effect for the remainder of the session, or until you remove it.

You are also right about serialisation, and probably we do *not* want to enable a system in which we use pickle or other more versatile system for config.

**Comment by elibixby** [comment]
@martindurant thanks for the swift response! That's a good idea for a slightly slicker way of solving the problem. I'd be a little worried about modifying a global dictionary given that fsspec could be used outside my code.

It's maybe a little paranoid but even if I wrote a context manager outside of fsspec I could accidentally call some external code that calls fsspec and uses the other default unintentionally, which is why it would be nice to have an additional argument to `open()` that uses a context that is scoped inside the open call.

**Comment by martindurant** [comment]
A context manager is a fine idea, but I would not add is as an argument to open, since there are other functions that create filesystems too. I would probably  make a simpler version of `dask.config.set` if I had the time.
